chainer/functions/connection/linear.py

```diff
@@ -45,26 +53,114 @@ def forward(self, inputs):
             x = numpy.ascontiguousarray(x)
 
         y = x.dot(W.T).astype(x.dtype, copy=False)
-        if len(inputs) == 3:
-            b = inputs[2]
+        if b is not None:
             y += b
         self.retain_inputs((0, 1))  # b is not retained
         return y,
 
+    def _forward_ideep(self, inputs):
+        if len(inputs) == 3:
+            x, W, b = inputs
+        else:
+            (x, W), b = inputs, None
+
+        y = intel64.ideep.linear.Forward(
+            intel64.ideep.array(x),
+            intel64.ideep.array(W),
+            intel64.ideep.array(b) if b is not None else None)
+
+        self.retain_inputs((0, 1))
+        return y,
+
     def backward(self, indexes, grad_outputs):
-        x, W = self.get_retained_inputs()
+        ret = []
         gy, = grad_outputs
 
+        x, W = self.get_retained_inputs()
+
+        if self._use_ideep:
+
+            # iDeep implementation
+            if 0 in indexes:  # grad_x
+                gx = LinearGradDIdeep().apply((W, gy))
+                ret.append(gx[0])
+            if 1 in indexes:  # grad_W
+                gW = LinearGradWIdeep().apply((x, gy))
+                ret.append(gW[0])
+            if 2 in indexes:  # grad_b
+                gb = chainer.functions.sum(gy, axis=0)
+                ret.append(gb)
+        else:
+            # Generic implementation
+            if 0 in indexes:
+                gx, = LinearGradData().apply((W, gy))
+                ret.append(chainer.functions.cast(gx, x.dtype))
+            if 1 in indexes:
+                gW, = LinearGradWeight().apply((x, gy))
+                ret.append(chainer.functions.cast(gW, W.dtype))
+            if 2 in indexes:
+                gb = chainer.functions.sum(gy, axis=0)
+                ret.append(gb)
+
+        return ret
+
+
+class LinearGradDIdeep(function_node.FunctionNode):
```

Comment:
Why are these backward function node classes separate from the original ones?

Reply:
That would require passing `use_ideep` flag to each grad class and make explicit `if-else` branches.
Rather than that, isn't it cleaner to separate the classes?

Reply:
Hmm. I'm concerning the maintenance cost; the implementation of the ideep version looks very similar to the original one (esp. the backward implementations are almost same). Can't we at least unify the backward implementations?

Reply:
OK, will do.